x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/TextCategorization.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.categorization;
+
+import org.apache.lucene.util.Accountable;
+import org.apache.lucene.util.RamUsageEstimator;
+
+import java.util.Arrays;
+
+import static org.elasticsearch.xpack.ml.aggs.categorization.CategorizationBytesRefHash.WILD_CARD_ID;
+
+/**
+ * A text categorization group that provides methods for:
+ *  - calculating similarity between it and a new log
+ *  - expanding the existing log group by adding a new log
+ */
+class TextCategorization implements Accountable {
+
+    private final long id;
+    private final int[] categorization;
+    private final long[] tokenCounts;
+    private long count;
+
+    // Used at the shard level for tracking the bucket ordinal for collecting sub aggregations
+    long bucketOrd;
+
+    TextCategorization(int[] logTokenIds, long count, long id) {
+        this.id = id;
+        this.categorization = logTokenIds;
+        this.count = count;
+        this.tokenCounts = new long[logTokenIds.length];
+        Arrays.fill(this.tokenCounts, count);
+    }
+
+    public long getId() {
+        return id;
+    }
+
+    int[] getCategorization() {
+        return categorization;
+    }
+
+    public long getCount() {
+        return count;
+    }
+
+    Similarity calculateSimilarity(int[] logEvent) {
+        assert logEvent.length == this.categorization.length;
+        int eqParams = 0;
+        long tokenCount = 0;
+        long tokensKept = 0;
+        for (int i = 0; i < logEvent.length; i++) {
+            if (logEvent[i] == this.categorization[i]) {
+                tokensKept += tokenCounts[i];
+                tokenCount += tokenCounts[i];
+            } else if (this.categorization[i] == WILD_CARD_ID) {
+                eqParams++;
+            } else {
+                tokenCount += tokenCounts[i];
+            }
+        }
+        return new Similarity((double) tokensKept / tokenCount, eqParams);
+    }
+
+    void addLog(int[] logEvent, long docCount) {
+        assert logEvent.length == this.categorization.length;
+        for (int i = 0; i < logEvent.length; i++) {
+            if (logEvent[i] != this.categorization[i]) {
+                this.categorization[i] = WILD_CARD_ID;
+            } else {
+                tokenCounts[i] += docCount;
+            }
+        }
+        this.count += docCount;
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        return Long.BYTES // id
+            + RamUsageEstimator.NUM_BYTES_OBJECT_REF // categorization reference
```

Comment:
This one too might benefit from a constant for the size of the instance. It'd be `BASE + RamUsageEstimator.sizeOf(categorization) + RamUsageEstimator.sizeOf(tokenCounts)`.